### PR TITLE
Add the ability to import a SARIF file (AST-36884, AST-36890)

### PR DESCRIPTION
**Pull request by checkmarx-kobi-hagmi**
### Testing
* Added unit and integration tests for the import command with different files.

### Checklist

- [ ] I have added documentation for new/changed functionality in this PR (if applicable).
- [x] I have updated the CLI help for new/changed functionality in this PR (if applicable).
- [x] All active GitHub checks for tests, formatting, and security are passing
- [x] The correct base branch is being used

### CODE REVIEW

File: internal/commands/import.go [hunk 1]
```diff
@@ -40,11 +41,34 @@ func runImportCommand(wrapper wrappers.UploadsWrapper) func(cmd *cobra.Command,
 		if err != nil {
 			return err
 		}
+		projectName, err := cmd.Flags().GetString(commonParams.ProjectName)
```
Comment (OrShamirCM) [inline internal/commands/import.go:44]: should we always have a project in this command? it seems like we had different functonaility

File: internal/commands/import.go [hunk 2]
```diff
@@ -40,11 +41,34 @@ func runImportCommand(wrapper wrappers.UploadsWrapper) func(cmd *cobra.Command,
...
+		if err != nil {
+			return err
+		}
 
 		if importFileType == "" || importFilePath == "" {
 			return errors.Errorf(clierrors.MissingImportFlags)
 		}
-		_, err = importFile(importFileType, importFilePath)
+		if projectName == "" {
```
Comment (OrShamirCM) [inline internal/commands/import.go:52]: same in here

File: internal/commands/import.go [hunk 3]
```diff
@@ -40,11 +41,34 @@ func runImportCommand(wrapper wrappers.UploadsWrapper) func(cmd *cobra.Command,
...
+			return errors.Errorf(clierrors.ProjectNameIsRequired)
+		}
+
+		params := make(map[string]string)
+		params["name"] = projectName
+		resp, _, err := projectsWrapper.Get(params)
```
Comment (OrShamirCM) [inline internal/commands/import.go:58]: and here. btw, all the project related - maybe you can export it to a private method

File: internal/commands/import.go [hunk 4]
```diff
@@ -53,7 +77,7 @@ func runImportCommand(wrapper wrappers.UploadsWrapper) func(cmd *cobra.Command,
 	}
 }
 
-func importFile(fileType string, path string) (string, error) {
+func importFile(projectId, fileType, path string) (string, error) {
```
Comment (OrShamirCM) [inline internal/commands/import.go:80]: if we don't need always projectid have a different importfile func / add int he last ...proiect id (to have default value nil)

File: internal/commands/import_test.go [hunk 5]
```diff
@@ -10,15 +10,20 @@ import (
 )
 
 func TestImport_ImportSarifFileWithCorrectFlags_CreateImportSuccessfully(t *testing.T) {
```
Comment (OrShamirCM) [inline internal/commands/import_test.go:12]: same here for project/ I may misunderstand the current functionality and the new

File: internal/commands/scan_test.go [hunk 6]
```diff
@@ -128,14 +128,18 @@ func TestScanCreate_ExistingApplicationAndProject_CreateProjectUnderApplicationS
 	execCmdNilAssertion(t, "scan", "create", "--project-name", "MOCK", "--application-name", "MOCK", "-s", dummyRepo, "-b", "dummy_branch")
 }
 
+func TestScanCreate_PassingImportFileTypeAndImportFilePath_CreateProjectAndImportSARIFFileSuccessfully(t *testing.T) {
+	execCmdNilAssertion(t, "scan", "create", "--project-name", "MOCK", "--import-file-path", "my-path", "--import-file-type", "my-type", "-s", dummyRepo, "-b", "dummy_branch")
```
Comment (OrShamirCM) [inline internal/commands/scan_test.go:132]: my-type should be SARIF ?

File: internal/commands/scan.go [hunk 7]
```diff
@@ -826,6 +828,24 @@ func setupScanTypeProjectAndConfig(
...
+		return err
+	}
+	importFilePath, err := cmd.Flags().GetString(commonParams.ImportFilePath)
+	if err != nil {
+		return err
+	}
+	if importFileType != "" && importFilePath != "" {
+		_, importFileError := importFile(projectID, importFileType, importFilePath)
+		if importFileError != nil {
+			logger.Printf(importFileError.Error())
```
Comment (OrShamirCM) [inline internal/commands/scan.go:843]: write your error here maybe / make sure we will not miss the error message we want to shoe the customer

**Review by tiagobcx — APPROVED** [review]

File: internal/commands/import.go (hunk 3 above)
Comment (checkmarx-kobi-hagmi) [reply internal/commands/import.go]: Done under the projectWrapper

File: internal/commands/import.go (hunk 1 above)
Comment (checkmarx-kobi-hagmi) [reply internal/commands/import.go:42]: import is the new command and project name must be provided

File: internal/commands/scan_test.go (hunk 6 above)
Comment (checkmarx-kobi-hagmi) [reply internal/commands/scan_test.go]: not relevant anymore. we removed this flag

File: internal/commands/scan.go (hunk 7 above)
Comment (checkmarx-kobi-hagmi) [reply internal/commands/scan.go]: the error we will show to the customer is mapped inside importFile, which will be implemented by Tamar

File: internal/commands/import.go [hunk 8]
```diff
@@ -9,42 +9,50 @@ import (
 	"github.com/spf13/cobra"
 )
 
-func NewImportCommand(uploadsWrapper wrappers.UploadsWrapper) *cobra.Command {
+func NewImportCommand(projectsWrapper wrappers.ProjectsWrapper, uploadsWrapper wrappers.UploadsWrapper) *cobra.Command {
```
Comment (OrShamirCM) [inline internal/commands/import.go:12]: Maybe hide it for now (cobra hide)

File: internal/commands/import_test.go [hunk 9]
```diff
@@ -10,15 +10,15 @@ import (
...
-	execCmdNilAssertion(t, "import", "--import-file-type", "my-type", "--import-file-path", "my-path")
+	execCmdNilAssertion(t, "import", "--project-name", "my-project", "--import-file-path", "my-path")
```
Comment (OrShamirCM) [inline internal/commands/import_test.go:13]: proejct not exsits test

File: internal/wrappers/projects-http.go [hunk 10]
```diff
@@ -238,3 +263,16 @@ func (p *ProjectsHTTPWrapper) Tags() (
 		return nil, nil, errors.Errorf("response status code %d", resp.StatusCode)
 	}
 }
+
+func setDefaultLimit(params map[string]string) {
```
Comment (OrShamirCM) [inline internal/wrappers/projects-http.go:267]: it's not clear.. good idea but it's need a deeper refactoring

File: internal/wrappers/projects-http.go [hunk 11]
```diff
@@ -148,6 +147,32 @@ func (p *ProjectsHTTPWrapper) GetByID(projectID string) (
...
+	*ProjectResponseModel,
+	*ErrorModel,
+	error) {
+	params := make(map[string]string)
+	params["name"] = name
+	resp, _, err := p.Get(params)
+	if err != nil {
+		return nil, nil, err
+	}
+	setLimit(params, "10")
```
Comment (OrShamirCM) [inline internal/wrappers/projects-http.go:160]: can be removed, it exists in the get call

File: internal/wrappers/projects-http.go [hunk 12]
```diff
@@ -148,6 +147,32 @@ func (p *ProjectsHTTPWrapper) GetByID(projectID string) (
...
+		return nil, nil, errors.Errorf(clierrors.ProjectNotExists)
+	}
+
+	for i := range resp.Projects {
+		if resp.Projects[i].Name == name {
+			return &resp.Projects[i], nil, nil
+		}
+	}
+
+	return nil, nil, errors.Errorf("The project name you provided does not match any project")
```
Comment (OrShamirCM) [inline internal/wrappers/projects-http.go:173]: does not exists

File: test/integration/project_test.go [hunk 13]
```diff
@@ -101,18 +100,22 @@ func TestProjectCreate_ApplicationDoesntExist_FailAndReturnErrorMessage(t *testi
 		flag(params.ApplicationName), "application-that-doesnt-exist",
 	)
```
Comment (OrShamirCM) [inline test/integration/project_test.go:102]: add test to get by name, with and without exact match

File: internal/commands/import_test.go (hunk 9 above)
Comment (checkmarx-kobi-hagmi) [reply internal/commands/import_test.go:13]: done

File: internal/commands/project.go [hunk 14]
```diff
@@ -217,7 +216,7 @@ func updateProjectRequestValues(input *[]byte, cmd *cobra.Command) error {
 	if projectName != "" {
 		info["name"] = projectName
 	} else {
-		return errors.Errorf("Project name is required")
+		return errors.Errorf(clierrors.ProjectNameIsRequired)
```
[the remark this answers is not in the document]
Comment (checkmarx-kobi-hagmi) [reply internal/commands/project.go:219]: done

File: test/integration/project_test.go (hunk 13 above)
Comment (checkmarx-kobi-hagmi) [reply test/integration/project_test.go:102]: irrelevant after last change

File: internal/commands/project.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (checkmarx-kobi-hagmi) [reply internal/commands/project.go]: Error: don't use MixedCaps in package name; errorConstants should be errorconstants (golint)
Also, cannot be changed to "errors" because of the standard errors library.
Changed to errorconsts

File: internal/wrappers/projects-http.go (hunk 12 above)
Comment (checkmarx-kobi-hagmi) [reply internal/wrappers/projects-http.go]: irrelevant after the change

File: internal/wrappers/projects-http.go (hunk 11 above)
Comment (checkmarx-kobi-hagmi) [reply internal/wrappers/projects-http.go]: done

File: internal/wrappers/projects-http.go (hunk 10 above)
Comment (checkmarx-kobi-hagmi) [reply internal/wrappers/projects-http.go]: done

File: internal/commands/import.go (hunk 8 above)
Comment (checkmarx-kobi-hagmi) [reply internal/commands/import.go]: hide what?

File: test/integration/import_test.go [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+	projectName := projectNameRandom
+
+	args := []string{
+		"import",
+		flag(params.ProjectName), projectName,
+		flag(params.ImportFilePath), "./data/sarif.sarif",
+	}
+	err, _ := executeCommand(t, args...)
+	assert.NilError(t, err, "import failed")
+}
```
Comment (checkmarx-kobi-hagmi) [inline test/integration/import_test.go:34]: Note: on a separate task, we should make it possible to delete a project by name

File: internal/commands/import_test.go [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+func TestImport_ImportSarifFileUnacceptedFileExtension_CreateImportReturnsErrorWithCorrectMessage(t *testing.T) {
+	err := execCmdNotNilAssertion(t, "import", "--project-name", "MOCK-PROJECT-NOT-EXIST", "--import-file-path", "my-path.txt")
+	assert.Assert(t, err.Error() == errorconstants.SarifInvalidFileExtension)
+}
+
+func TestImport_ImportSarifFileMissingExtension_CreateImportReturnsErrorWithCorrectMessage(t *testing.T) {
+	err := execCmdNotNilAssertion(t, "import", "--project-name", "MOCK-PROJECT-NOT-EXIST", "--import-file-path", "some/path/no/extension/my-path")
+	assert.Assert(t, err.Error() == errorconstants.SarifInvalidFileExtension)
+}
+
```
Comment (checkmarx-kobi-hagmi) [inline internal/commands/import_test.go:45]: The following tests for "importFile" were created to test the mapping of the http error codes to user-friendly error codes.
When we will be able to define an expectation of mocks we'll be able to properly test the wrappers by mocking client.go

File: internal/commands/root.go [hunk 17]
```diff
@@ -162,6 +163,9 @@ func NewAstCLI(
 		accessManagementWrapper,
 	)
 	projectCmd := NewProjectCommand(applicationsWrapper, projectsWrapper, groupsWrapper, accessManagementWrapper)
+
+	importCmd := NewImportCommand(projectsWrapper, uploadsWrapper, groupsWrapper, accessManagementWrapper, byorWrapper)
```
Comment (OrShamirCM) [inline internal/commands/root.go:167]: please check the FF (check which FF wirh Edurate team)

File: internal/commands/scan.go [hunk 18]
```diff
@@ -18,11 +18,11 @@ import (
 	"strings"
 	"time"
 
-	applicationErrors "github.com/checkmarx/ast-cli/internal/errors"
-
 	"github.com/checkmarx/ast-cli/internal/commands/scarealtime"
 	"github.com/checkmarx/ast-cli/internal/commands/util"
 	"github.com/checkmarx/ast-cli/internal/commands/util/printer"
+	"github.com/checkmarx/ast-cli/internal/constants"
+	errorconstants "github.com/checkmarx/ast-cli/internal/constants/errors"
```
Comment (OrShamirCM) [inline internal/commands/scan.go:25]:
```suggestion
	errorConstants "github.com/checkmarx/ast-cli/internal/constants/errors"
```

File: test/integration/import_test.go (hunk 15 above)
Comment (OrShamirCM) [reply test/integration/import_test.go:34]: why on different? we already had a general cleanup test, let's have a cleanup here as well

Comment (OrShamirCM) [reply test/integration/import_test.go:34]: you can do it by get and delete

Comment (OrShamirCM) [reply test/integration/import_test.go:34]: BTW, if the project doesn't exist we agreed on creating new one. so why did this test assert fail?

File: test/integration/scan_test.go [hunk 19]
```diff
@@ -22,7 +22,6 @@ import (
 	realtime "github.com/checkmarx/ast-cli/internal/commands/scarealtime"
 	"github.com/checkmarx/ast-cli/internal/commands/util"
 	"github.com/checkmarx/ast-cli/internal/commands/util/printer"
-	applicationErrors "github.com/checkmarx/ast-cli/internal/errors"
```
Comment (OrShamirCM) [inline test/integration/scan_test.go:25]: you removed this import, this why the linter have issues

File: .golangci.yml [hunk 20]
```diff
@@ -12,6 +12,7 @@ linters-settings:
           - github.com/spf13/cobra
           - github.com/pkg/errors
           - github.com/google
+          - github.com/MakeNowJust/heredoc
```
Comment (OrShamirCM) [inline .golangci.yml:15]: why did you add this packge ?

File: test/integration/scan_test.go (hunk 19 above)
Comment (checkmarx-kobi-hagmi) [reply test/integration/scan_test.go:25]: Fixed

File: test/integration/import_test.go (hunk 15 above)
Comment (checkmarx-kobi-hagmi) [reply test/integration/import_test.go:86]: Done

File: internal/commands/root.go (hunk 17 above)
Comment (checkmarx-kobi-hagmi) [reply internal/commands/root.go:167]: Discussed at office

File: internal/wrappers/feature-flags.go [hunk 21]
```diff
@@ -26,6 +26,15 @@ var FeatureFlagsBaseMap = []CommandFlags{
 	{
 		CommandName: "cx project create",
 	},
+	{
+		CommandName: "cx import",
+		FeatureFlags: []FlagBase{
```
Comment (tamarleviCm) [inline internal/wrappers/feature-flags.go:31]: you don't need to set up the FeatureFlags list.
if you use a new FF in your code you have to add it here with default value.
if you don't use the FF at all - you don't need all this item of "cx import"

File: internal/wrappers/projects-http.go [hunk 22]
```diff
@@ -148,6 +149,31 @@ func (p *ProjectsHTTPWrapper) GetByID(projectID string) (
 	return handleProjectResponseWithBody(resp, err, http.StatusOK)
 }
 
+func (p *ProjectsHTTPWrapper) GetByName(name string) (
```
Comment (tamarleviCm) [inline internal/wrappers/projects-http.go:152]: you didn't use this func on code. please remove it

File: internal/wrappers/projects.go [hunk 23]
```diff
@@ -51,6 +51,7 @@ type ProjectsWrapper interface {
 	Update(projectID string, model *Project) error
 	Get(params map[string]string) (*ProjectsCollectionResponseModel, *ErrorModel, error)
 	GetByID(projectID string) (*ProjectResponseModel, *ErrorModel, error)
+	GetByName(name string) (*ProjectResponseModel, *ErrorModel, error)
```
Comment (tamarleviCm) [inline internal/wrappers/projects.go:54]: you didn't use this func from the code

File: test/integration/import_test.go [hunk 24]
```diff
@@ -0,0 +1,78 @@
...
+	"testing"
+
+	errorConstants "github.com/checkmarx/ast-cli/internal/constants/errors"
+	"github.com/checkmarx/ast-cli/internal/params"
+	"gotest.tools/assert"
+)
+
+func TestImport_ImportSarifFileWithCorrectFlags_CreateImportSuccessfully(t *testing.T) {
+	projectId, projectName := createProject(t, nil, nil)
+	defer deleteProject(t, projectId)
```
Comment (tamarleviCm) [inline test/integration/import_test.go:15]: isn't it duplicated with line 22

File: test/integration/import_test.go [hunk 25]
```diff
@@ -0,0 +1,78 @@
...
+		flag(params.ImportFilePath), ".\\data\\sarif.sarif",
+	}
+	err, _ := executeCommand(t, args...)
+	defer deleteProjectByName(t, projectName)
+	assert.NilError(t, err, "Import command failed with existing project")
+}
+
+func TestImport_ImportSarifFileWithCorrectFlagsZipFileExtention_CreateImportSuccessfully(t *testing.T) {
+	projectId, projectName := createProject(t, nil, nil)
+	defer deleteProject(t, projectId)
```
Comment (tamarleviCm) [inline test/integration/import_test.go:28]: isn't it duplicated with line 35

File: internal/commands/import_test.go [hunk 26]
```diff
@@ -0,0 +1,59 @@
...
+	assert.Assert(t, err.Error() == errorConstants.SarifInvalidFileExtension)
+}
+
+func TestImport_ImportSarifFileMissingExtension_CreateImportReturnsErrorWithCorrectMessage(t *testing.T) {
+	err := execCmdNotNilAssertion(t, "import", "--project-name", "MOCK-PROJECT-NOT-EXIST", "--import-file-path", "some/path/no/extension/my-path")
+	assert.Assert(t, err.Error() == errorConstants.SarifInvalidFileExtension)
+}
+
+func TestImporFileFunction_FakeUnauthorizedHttpStatusCode_ReturnRelevantError(t *testing.T) {
+	err := importFile(mock.FakeUnauthorized401, "importFilePath", &mock.UploadsMockWrapper{}, &mock.ByorMockWrapper{})
```
[the remark this answers is not in the document]
Comment (checkmarx-kobi-hagmi) [reply internal/commands/import_test.go:47]: Because it's not a command, but testing of the importFile function. These importFile tests are meant to verify that the http errors are mapped to user friendly errors.

File: internal/wrappers/projects-http.go (hunk 22 above)
Comment (checkmarx-kobi-hagmi) [reply internal/wrappers/projects-http.go:152]: It is used in integration tests

File: test/integration/import_test.go (hunk 24 above)
Comment (checkmarx-kobi-hagmi) [reply test/integration/import_test.go]: Fixed

File: internal/wrappers/projects.go (hunk 23 above)
Comment (checkmarx-kobi-hagmi) [reply internal/wrappers/projects.go:54]: It is used by "deleteProjectByName"

File: internal/commands/root.go [hunk 27]
```diff
@@ -205,6 +210,10 @@ func NewAstCLI(
 		chatCmd,
 	)
 
+	if wrappers.FeatureFlags[featureFlagsConstants.BYOR_ENABLED] {
```
Comment (OrShamirCM) [inline internal/commands/root.go:213]: please set default valu + test it

File: internal/wrappers/projects.go (hunk 23 above)
Comment (tamarleviCm) [reply internal/wrappers/projects.go:54]: we don't add code only for tests (only if it mast mast)

File: internal/commands/import.go [hunk 28]
```diff
@@ -0,0 +1,111 @@
...
+
+		projectName, err := getProjectName(cmd)
+		if err != nil {
+			return err
+		}
+
+		projectID, err := findProject(nil, projectName, cmd, projectsWrapper, groupsWrapper, accessManagementWrapper)
+		if err != nil {
+			return err
+		}
```
[the remark this answers is not in the document]
Comment (checkmarx-kobi-hagmi) [reply internal/commands/import.go:62]: Done

File: internal/wrappers/feature-flags.go (hunk 21 above)
Comment (checkmarx-kobi-hagmi) [reply internal/wrappers/feature-flags.go:32]: This item must be there because it's used by the uploadsWrapper, used by the import command..

File: internal/commands/root.go (hunk 27 above)
Comment (checkmarx-kobi-hagmi) [reply internal/commands/root.go]: Done

File: internal/commands/root.go [hunk 29]
```diff
@@ -205,6 +210,10 @@ func NewAstCLI(
 		chatCmd,
 	)
 
+	if wrappers.FeatureFlags[featureFlagsConstants.ByorEnabled] {
```
Comment (tamarleviCm) [inline internal/commands/root.go:213]: I see you use the FF now,
few points to pay attention:
1. do you want to notify the user that this command is not available because the FF is off? if so, you need to add the command and hide him.
2. you have to add a test to check the situation with FF off
3. in the unit tests you have to set the FF=on - Otherwise the tests will fail (you can take reference from groups_test.go) 
4. you have to verify that the integration test will not fail when the FF is off

File: internal/commands/root.go (hunk 27 above)
Comment (OrShamirCM) [reply internal/commands/root.go]: make sure that it happen (manualy), add be changes/after ben merge

File: test/integration/import_test.go [hunk 30]
```diff
@@ -6,61 +6,80 @@ import (
...
 		flag(params.ImportFilePath), ".\\data\\sarif.sarif",
 	}
 	err, _ := executeCommand(t, args...)
-	defer deleteProjectByName(t, projectName)
+
 	assert.NilError(t, err, "Import command failed with non-existing project")
 }
 
 func TestImport_ImportSarifFileMissingVersion_ImportFailWithCorrectMessage(t *testing.T) {
+	t.Skip("BYOR flag is currently disabled")
```
Comment (OrShamirCM) [inline test/integration/import_test.go:67]: remove skip to be by FF

**Review by tamarleviCm — APPROVED** [review]

**Review by OrShamirCM — CHANGES_REQUESTED** [review]
Update branch to master, have e2e tests with UI and because there are a lot of change in the source doce we need regression tests for creating / update project with without groups and access management + scan

File: internal/commands/groups.go [hunk 31]
```diff
@@ -2,110 +2,25 @@ package commands
...
 	commonParams "github.com/checkmarx/ast-cli/internal/params"
+	"github.com/checkmarx/ast-cli/internal/services"
 	"github.com/checkmarx/ast-cli/internal/wrappers"
-	"github.com/pkg/errors"
 	"github.com/spf13/cobra"
 )
 
-const accessManagementEnabled = "ACCESS_MANAGEMENT_ENABLED" // feature flag
-
-func createGroupsMap(groupsStr string, groupsWrapper wrappers.GroupsWrapper) ([]*wrappers.Group, error) {
```
Comment (OrShamirCM) [inline internal/commands/groups.go:15]: all groups related logic suppose to be in a group service and not under project service, we need to keep the same separation

Comment (OrShamirCM) [reply internal/commands/groups.go:15]: maybe all this group file suppose to be a "service" or a domain inside project

Comment (checkmarx-kobi-hagmi) [reply internal/commands/groups.go:15]: done

**Comment by OrShamirCM** [comment]
> Update branch to master, have e2e tests with UI and because there are a lot of change in the source doce we need regression tests for creating / update project with without groups and access management + scan

I see you did the separation, great! if you can create test file for each of them and move the tests to there it'll be great.

File: internal/services/applications_test.go [hunk 32]
```diff
@@ -0,0 +1,39 @@
...
+	type args struct {
+		applicationID          []string
+		existingApplicationIds []string
+	}
+	tests := []struct {
+		name string
+		args args
+		want []string
+	}{
+		// TODO: Add test cases.
```
Comment (OrShamirCM) [inline internal/services/applications_test.go:18]: What is this TODO: ?
Its is still relevant?

File: internal/services/projects_test.go [hunk 33]
```diff
@@ -0,0 +1,281 @@
...
+		accessManagementWrapper wrappers.AccessManagementWrapper
+		applicationsWrapper     wrappers.ApplicationsWrapper
+	}
+	tests := []struct {
+		name    string
+		args    args
+		want    string
+		wantErr bool
+	}{
+		// TODO: Add test cases.
```
Comment (OrShamirCM) [inline internal/services/projects_test.go:27]: same here

File: internal/services/projects_test.go [hunk 34]
```diff
@@ -0,0 +1,281 @@
...
+		projectGroups           string
+		projectPrivatePackage   string
+	}
+	tests := []struct {
+		name    string
+		args    args
+		want    string
+		wantErr bool
+	}{
+		// TODO: Add test cases.
```
Comment (OrShamirCM) [inline internal/services/projects_test.go:96]: same here

File: internal/services/projects_test.go [hunk 35]
```diff
@@ -0,0 +1,281 @@
...
+		projectTags             string
+		projectPrivatePackage   string
+	}
+	tests := []struct {
+		name    string
+		args    args
+		want    string
+		wantErr bool
+	}{
+		// TODO: Add test cases.
```
Comment (OrShamirCM) [inline internal/services/projects_test.go:187]: and here

File: internal/services/tags_test.go [hunk 36]
```diff
@@ -0,0 +1,32 @@
...
+func Test_createTagMap(t *testing.T) {
+	type args struct {
+		tagListStr string
+	}
+	tests := []struct {
+		name string
+		args args
+		want map[string]string
+	}{
+		// TODO: Add test cases.
```
Comment (OrShamirCM) [inline internal/services/tags_test.go:17]: and here

File: internal/wrappers/byor-http.go [hunk 37]
```diff
@@ -0,0 +1,78 @@
...
+		_ = resp.Body.Close()
+	}()
+	decoder := json.NewDecoder(resp.Body)
+	switch resp.StatusCode {
+	case http.StatusForbidden:
+		return "", getError(decoder, errorConstants.StatusForbidden)
+	case http.StatusUnauthorized:
+		return "", getError(decoder, errorConstants.StatusUnauthorized)
+	case http.StatusInternalServerError:
+		return "", getError(decoder, errorConstants.StatusInternalServerError)
```
Comment (OrShamirCM) [inline internal/wrappers/byor-http.go:53]: can it be  “There was a problem importing the SARIF file. Please contact support for further detail with the following error code: “xxxx + API MESSAGE” instead if internal status error?
we can so it in different task if it's complecated

File: internal/wrappers/byor-http.go [hunk 38]
```diff
@@ -0,0 +1,78 @@
...
+		model := CreateImportsResponse{}
+		err = decoder.Decode(&model)
+		if err != nil {
+			return model.ImportID, errors.Errorf("Parsing upload model failed - %s", err.Error())
+		}
+		logger.Printf(successfulMessage, projectID, model.ImportID)
+		return model.ImportID, nil
+	case http.StatusNotFound:
+		fmt.Printf("status code: %d b.path+importsPath: %s request: %s\n", http.StatusNotFound, b.path+importsPath, req)
+		return "", errors.Errorf("Import was not successful for project ID %s", projectID)
```
Comment (OrShamirCM) [inline internal/wrappers/byor-http.go:64]: when are we getting this 404? why should we get it? is it sever error of project not exits? if project not exsits, let's response this?

File: internal/wrappers/byor-http.go [hunk 39]
```diff
@@ -0,0 +1,78 @@
...
+	default:
+		return "", errors.Errorf(errorConstants.ImportSarifFileErrorMessage)
```
Comment (OrShamirCM) [inline internal/wrappers/byor-http.go:66]: if defatult can be -  “There was a problem importing the SARIF file. Please contact support for further detail with the following error code: “xxxx + API MESSAGE”

File: internal/services/projects_test.go (hunk 33 above)
Comment (checkmarx-kobi-hagmi) [reply internal/services/projects_test.go:27]: It was auto-generated comment Goland to guide the developer will see where to add the test data.
I'll remove it.

File: internal/wrappers/byor-http.go (hunk 37 above)
Comment (checkmarx-kobi-hagmi) [reply internal/wrappers/byor-http.go:53]: It's not complicated, but maybe it's also a good idea to decide on a pattern that will be used in all of out http wrappers when we get error 500?

File: internal/wrappers/byor-http.go (hunk 38 above)
Comment (checkmarx-kobi-hagmi) [reply internal/wrappers/byor-http.go:64]: I think I got this one when there was some temporary issue with the byor service and it returned 404.
It sends the upoad url it gets from the the Upload service, which was generated successfully.

Comment (checkmarx-kobi-hagmi) [reply internal/wrappers/byor-http.go:64]: Removed

File: internal/wrappers/byor-http.go (hunk 37 above)
Comment (checkmarx-kobi-hagmi) [reply internal/wrappers/byor-http.go]: Fixed

File: internal/wrappers/byor-http.go (hunk 39 above)
Comment (checkmarx-kobi-hagmi) [reply internal/wrappers/byor-http.go]: in theory this code should not be reached. from looking in other wrappers, they don't use ErrorModel as used in other error status codes, so I return an general error. example from project-http:
	default:
		return nil, nil, errors.Errorf("response status code %d", resp.StatusCode)

File: internal/wrappers/byor-http.go (hunk 38 above)
Comment (OrShamirCM) [reply internal/wrappers/byor-http.go]: please commit

File: internal/wrappers/byor-http.go [hunk 40]
```diff
@@ -50,7 +50,12 @@ func (b *ByorHTTPWrapper) Import(projectID, uploadURL string) (string, error) {
 	case http.StatusUnauthorized:
 		return "", getError(decoder, errorConstants.StatusUnauthorized)
 	case http.StatusInternalServerError:
-		return "", getError(decoder, errorConstants.StatusInternalServerError)
+		errorModel := ErrorModel{}
```
Comment (OrShamirCM) [inline internal/wrappers/byor-http.go:53]: veirfy byor error model - https://deu.ast.checkmarx.net/spec/v1?urls.primaryName=Byor%20Results%20Handler#/default/post_imports

File: internal/wrappers/byor-http.go [hunk 41]
```diff
@@ -50,7 +50,12 @@ func (b *ByorHTTPWrapper) Import(projectID, uploadURL string) (string, error) {
...
+		decodeErr := decoder.Decode(&errorModel)
+		if decodeErr != nil {
+			return "", errors.Errorf("Parsing error model failed - %s", err.Error())
```
Comment (OrShamirCM) [inline internal/wrappers/byor-http.go:56]: return the const of error in byor but without code and message

File: internal/constants/errors/errors.go [hunk 42]
```diff
@@ -16,5 +16,5 @@ const (
 	ProjectNotExists                     = "the project name you provided does not match any project"
 	FailedToGetApplication               = "failed to get application"
 	SarifInvalidFileExtension            = "Invalid file extension. Supported extensions are .sarif and .zip containing sarif files."
-	ImportSarifFileErrorMessage          = "There was a problem importing the SARIF file"
+	ImportSarifFileErrorMessage          = "There was a problem importing the SARIF file. Please contact support for further detail with the following error code: %d %s"
```
Comment (OrShamirCM) [inline internal/constants/errors/errors.go]: maybe the error code stirng and mesage remove from const and have it dynamicly

File: internal/wrappers/byor-http.go [hunk 43]
```diff
@@ -59,11 +64,8 @@ func (b *ByorHTTPWrapper) Import(projectID, uploadURL string) (string, error) {
 		}
 		logger.Printf(successfulMessage, projectID, model.ImportID)
 		return model.ImportID, nil
-	case http.StatusNotFound:
-		fmt.Printf("status code: %d b.path+importsPath: %s request: %s\n", http.StatusNotFound, b.path+importsPath, req)
-		return "", errors.Errorf("Import was not successful for project ID %s", projectID)
 	default:
-		return "", errors.Errorf(errorConstants.ImportSarifFileErrorMessage)
+		return "", errors.Errorf("response status code %d", resp.StatusCode)
```
Comment (OrShamirCM) [inline internal/wrappers/byor-http.go:68]: return the const with error code (contatc support..)

File: internal/wrappers/byor-http.go (hunk 40 above)
Comment (checkmarx-kobi-hagmi) [reply internal/wrappers/byor-http.go:53]: fixed

File: internal/wrappers/byor-http.go (hunk 41 above)
Comment (checkmarx-kobi-hagmi) [reply internal/wrappers/byor-http.go:56]: fixed

File: internal/constants/errors/errors.go (hunk 42 above)
Comment (checkmarx-kobi-hagmi) [reply internal/constants/errors/errors.go]: fixed

File: internal/wrappers/byor-http.go (hunk 41 above)
Comment (OrShamirCM) [reply internal/wrappers/byor-http.go:56]: seems like not fixed yet?

File: internal/wrappers/byor-http.go [hunk 44]
```diff
@@ -0,0 +1,86 @@
...
+	case http.StatusOK:
+		model := CreateImportsResponse{}
+		err = decoder.Decode(&model)
+		if err != nil {
+			return model.ImportID, errors.Errorf("Parsing upload model failed - %s", err.Error())
+		}
+		logger.Printf(successfulMessage, projectID, model.ImportID)
+		return model.ImportID, nil
+	default:
+		return "", errors.Errorf(errorConstants.ImportSarifFileError)
```
Comment (OrShamirCM) [inline internal/wrappers/byor-http.go:68]: we desiced to return here ImportSarifFileErrorMessageWithMessage too with statuscode but no message

**Review by OrShamirCM — APPROVED** [review]
